crates/sdk/src/codec.rs

```diff
@@ -0,0 +1,544 @@
+use std::io::{self, Cursor, Read, Result, Write};
+
+use openvm_native_compiler::ir::DIGEST_SIZE;
+use openvm_native_recursion::hints::{InnerBatchOpening, InnerFriProof, InnerQueryProof};
+use openvm_stark_backend::{
+    config::{Com, PcsProof},
+    interaction::{fri_log_up::FriLogUpPartialProof, RapPhaseSeqKind},
+    p3_field::{
+        extension::BinomialExtensionField, FieldAlgebra, FieldExtensionAlgebra, PrimeField32,
+    },
+    proof::{AdjacentOpenedValues, AirProofData, Commitments, OpenedValues, OpeningProof, Proof},
+};
+use p3_fri::CommitPhaseProofStep;
+
+use super::{F, SC}; // BabyBearPoseidon2Config
+
+type Challenge = BinomialExtensionField<F, 4>;
+
+/// Codec version should change only when proof system or proof format changes.
+/// It does correspond to the main openvm version (which may change more frequently).
+const CODEC_VERSION: u32 = 1;
+
+/// Hardware and language independent encoding.
+/// Uses the Writer pattern for more efficient encoding without intermediate buffers.
+// @dev Trait just for implementation sanity
+pub trait Encode {
+    /// Writes the encoded representation of `self` to the given writer.
+    fn encode<W: Write>(&self, writer: &mut W) -> Result<()>;
+
+    /// Convenience method to encode into a Vec<u8>
+    fn encode_to_vec(&self) -> Result<Vec<u8>> {
+        let mut buffer = Vec::new();
+        self.encode(&mut buffer)?;
+        Ok(buffer)
+    }
+}
+
+/// Hardware and language independent decoding.
+/// Uses the Reader pattern for efficient decoding.
+pub trait Decode: Sized {
+    /// Reads and decodes a value from the given reader.
+    fn decode<R: Read>(reader: &mut R) -> Result<Self>;
+}
+
+// General Note [jpw]: even though `SC` is a concrete type below, for typedefs that rely on associated types such as PcsProof<SC>,
+// Rust will prevent you from implementing Encode on the typedef, (I think) because if the SC trait changes, then the associated
+// type may change. In these cases, either impl Encode on the concrete type, or make a separate `encode_something` function.
+
+pub fn encode_proof_to_bytes(proof: &Proof<SC>) -> Result<Vec<u8>> {
+    let mut buffer = Vec::new();
+    encode_proof(proof, &mut buffer)?;
+    Ok(buffer)
+}
+
+pub fn decode_proof_from_bytes(bytes: &[u8]) -> Result<Proof<SC>> {
+    let mut reader = Cursor::new(bytes);
+    decode_proof(&mut reader)
+}
+
+// ==================== Encode implementation ====================
+
+// We need to know:
+// - Pcs is TwoAdicFriPcs
+// - Com<SC>: Into<[F; 8]>
+// For simplicity, we only implement for fixed `BabyBearPoseidon2Config`
+//
+/// Encode a proof using FRI as the PCS with `BabyBearPoseidon2Config`.
+/// The Merkle tree hashes have digest `[F; 8]`.
```

Comment:
```suggestion
/// The Merkle tree hashes have digest `[F; 8]`.
/// ```
/// pub struct Proof<SC: StarkGenericConfig> {
///     pub commitments: Commitments<Com<SC>>,
///     pub opening: OpeningProof<PcsProof<SC>, SC::Challenge>,
///     pub per_air: Vec<AirProofData<Val<SC>, SC::Challenge>>,
///     pub rap_phase_seq_proof: Option<RapPhaseSeqPartialProof<SC>>,
/// }
/// ```
```